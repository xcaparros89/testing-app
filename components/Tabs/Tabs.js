import { useState, useEffect } from "react";
import styles from "./Tabs.module.css";
import { slugify } from "../../utils/slugify";
import PropTypes from "prop-types";
import { useRouter } from "next/router";

export const Tabs = ({ children, initialTab }) => {
  const [activeTab, setActiveTab] = useState(slugify(children[0].props.label));
  const router = useRouter();
  const handleClick = (e, newActiveTab) => {
    e.preventDefault();
    setActiveTab(slugify(newActiveTab));
  };
  useEffect(() => {
    if (initialTab.tab) {
      setActiveTab(initialTab.tab);
    }
  }, []);
  useEffect(() => {
    router.push(`${router.pathname}?tab=${slugify(activeTab)}`, undefined, {
      shallow: true,
    });
  }, [activeTab]);
  return (
    <div>
      <ul className={styles.tabs}>
        {children.map((tab, i) => {
          const label = tab.props.label;
          return (
            <li
              data-testid={slugify(label)}
              className={slugify(label) === activeTab ? styles.current : ""}
              key={i}
            >
              <a href="#" onClick={(e) => handleClick(e, label)}>
                {label}
              </a>
            </li>
          );
        })}
      </ul>
      {children.map((child, i) => {
        if (slugify(child.props.label) === activeTab) {
          return (
            <div data-testid="content" key={i} className={styles.content}>
              {child.props.children}
            </div>
          );
        }
      })}
    </div>
  );
};

Tabs.propTypes = {
  initialTab: PropTypes.object,
  children: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string,
      children: PropTypes.any,
    })
  ).isRequired,
};
Tabs.defaultProps = {
  initialTab: {},
};
