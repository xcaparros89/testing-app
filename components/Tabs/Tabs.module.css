.tabs{
    list-style:none;
    padding: 0;
    margin:0;
    display: flex;
    border-bottom: 2px solid blue;
}

.tabs > li{
    margin: 0 1em;
    padding: 0.5em 1em;
}

.current{
    color: blue;
    font-weight: bold;
    border: 2px solid blue;
    border-bottom: 0;
    border-top-right-radius: 5px;
    border-top-left-radius: 5px;
    background: white;
    position: relative;
    top: 2px;
}

.content{
    padding: 1em 2em
}