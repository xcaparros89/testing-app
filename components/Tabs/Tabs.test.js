import React from "react";
import ReactDOM from "react-dom";
import { render, fireEvent } from "@testing-library/react";
import '@testing-library/jest-dom'
import { Tabs } from "./Tabs";
import { useRouter } from "next/router";

jest.mock("next/router");
beforeEach(()=>{
    useRouter.mockReturnValue({
        push: () => {},
      });
})
test("<Tabs /> matches snapshot", () => {
  const component = render(
    <Tabs initialTab={{ tab: "tab-3" }}>
      <div label="Tab 1">
        <h2>Title 1</h2>
        <p>Text 1</p>
      </div>
      <div label="Tab 2">
        <h2>Title 2</h2>
        <p>Text 2</p>
      </div>
      <div label="Tab 3">
        <h2>Title 3</h2>
        <p>Text 3</p>
      </div>
    </Tabs>
  );
  expect(component.container).toMatchSnapshot();
});

test("<Tabs /> renders without crashing", () => {
  const div = document.createElement("div");
  render(
    <Tabs initialTab={{ tab: "tab-3" }}>
      <div label="Tab 1">
        <h2>Title 1</h2>
        <p>Text 1</p>
      </div>
      <div label="Tab 2">
        <h2>Title 2</h2>
        <p>Text 2</p>
      </div>
      <div label="Tab 3">
        <h2>Title 3</h2>
        <p>Text 3</p>
      </div>
    </Tabs>,
    div
  );
});
test("Defaults to the first tab's content", () => {
  const div = document.createElement("div");
  const {getByTestId} = render(
    <Tabs>
      <div label="Tab 1">
        1
      </div>
      <div label="Tab 2">
        2
      </div>
      <div label="Tab 3">
        3
      </div>
    </Tabs>,
    div
  );
  const content = getByTestId("content");
  expect(content.textContent).toBe('1')
});
test("Can set a different tab as the initial state", () => {
    const div = document.createElement("div");
    const {getByTestId} = render(
      <Tabs initialTab={{ tab: "tab-3" }}>
        <div label="Tab 1">
          1
        </div>
        <div label="Tab 2">
          2
        </div>
        <div label="Tab 3">
          3
        </div>
      </Tabs>,
      div
    );
    const content = getByTestId("content");
    expect(content.textContent).toBe('3')
});
test("Click from one tab to the next", () => {
    const div = document.createElement("div");
    const {getByTestId, getByText} = render(
      <Tabs initialTab={{ tab: "tab-3" }}>
        <div label="Tab 1">
          1
        </div>
        <div label="Tab 4">
          4
        </div>
        <div label="Tab 3">
          3
        </div>
      </Tabs>,
      div
    );
    let content = getByTestId("content");
    expect(content.textContent).toBe('3')
    //differentiate the current tab
    const li = getByTestId('tab-3');
    expect(li).toHaveClass('current');
    //click on a new tab
    const differentTab = getByText('Tab 4');
    fireEvent.click(differentTab);
    const diffLi = getByTestId('tab-4');
    expect(diffLi).toHaveClass('current');
    content = getByTestId("content");
    expect(content.textContent).toBe('4')
});
