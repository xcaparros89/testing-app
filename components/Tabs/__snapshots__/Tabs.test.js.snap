// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tabs /> matches snapshot 1`] = `
<div>
  <div>
    <ul
      class="tabs"
    >
      <li
        class=""
        data-testid="tab-1"
      >
        <a
          href="#"
        >
          Tab 1
        </a>
      </li>
      <li
        class=""
        data-testid="tab-2"
      >
        <a
          href="#"
        >
          Tab 2
        </a>
      </li>
      <li
        class="current"
        data-testid="tab-3"
      >
        <a
          href="#"
        >
          Tab 3
        </a>
      </li>
    </ul>
    <div
      class="content"
      data-testid="content"
    >
      <h2>
        Title 3
      </h2>
      <p>
        Text 3
      </p>
    </div>
  </div>
</div>
`;
